modules/gdscript/gdscript_parser.cpp

```diff
@@ -4274,12 +4301,23 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
 		if (!export_type.is_variant() && (export_type.kind != DataType::BUILTIN || export_type.builtin_type != t_type)) {
 			// Allow float/int conversion.
 			if ((t_type != Variant::FLOAT || export_type.builtin_type != Variant::INT) && (t_type != Variant::INT || export_type.builtin_type != Variant::FLOAT)) {
-				push_error(vformat(R"("%s" annotation requires a variable of type "%s" but type "%s" was given instead.)", p_annotation->name.operator String(), Variant::get_type_name(t_type), export_type.to_string()), variable);
+				Vector<Variant::Type> expected_types = { t_type };
+				push_error(_get_annotation_error_string(p_annotation->name, expected_types, variable->get_datatype()), variable);
 				return false;
 			}
 		}
 	}
 
+	if (is_array) {
+		String hint_prefix = itos(variable->export_info.type);
+		if (variable->export_info.hint) {
+			hint_prefix += "/" + itos(variable->export_info.hint);
+		}
+		variable->export_info.hint = PROPERTY_HINT_TYPE_STRING;
+		variable->export_info.hint_string = hint_prefix + ":" + variable->export_info.hint_string;
+		variable->export_info.type = Variant::ARRAY;
```

Comment:
```suggestion
		variable->export_info.type = variable->datatype.builtin_type;
```

This is needed for weak types. For them, the compiler uses export type:

https://github.com/godotengine/godot/blob/13ba673c42951fd7cfa6fd8a7f25ede7e9ad92bb/modules/gdscript/gdscript_compiler.cpp#L2746-L2755

You mostly used hard types, which is why we didn't notice the bug before.

Also please add some tests for weak types, like:

```gdscript
var temp_packed_int: PackedInt64Array
@export var test_packed_int = temp_packed_int
```

I think there is no need to list all the annotations.

Reply:
Hmm. Figured out how to fix it. I forgot to add a check for `is_array` in the `@export` variant check branch.

Reply:
Oh, good catch! In the case of `@export`, the nested type should not be unwraped at all. Added a suggestion.

Reply:
The solution was adding

`export_type.type_source = variable->datatype.type_source;`
in the packed array branch.